lib/seeing_is_believing/evaluate_by_moving_files.rb

```diff
@@ -83,68 +87,72 @@ def set_back_to_initial_conditions
     end
 
     def write_program_to_file
-      File.open(filename, 'w') { |f| f.write program.to_s }
+      File.open(filename, 'w', external_encoding: "utf-8") { |f| f.write program.to_s }
     end
 
     # have to basically copy a bunch of Open3 code into here b/c keywords don't work right when the keys are not symbols
     # https://github.com/ruby/ruby/pull/808    my PR
     # https://bugs.ruby-lang.org/issues/10699  they opened an issue
     # https://bugs.ruby-lang.org/issues/10118  weird feature vs bug conversation
     def evaluate_file
+      event_server = TCPServer.new(port)
+
       # setup streams
-      eventstream, child_eventstream = IO.pipe
-      stdout,      child_stdout      = IO.pipe
-      stderr,      child_stderr      = IO.pipe
-      child_stdin, stdin             = IO.pipe
+      stdout,      child_stdout      = IO.pipe("utf-8")
+      stderr,      child_stderr      = IO.pipe("utf-8")
 
       # setup environment variables
       env = ENV.to_hash.merge 'SIB_VARIABLES.MARSHAL.B64' =>
                                 [Marshal.dump(
-                                  event_stream_fd:   4,
+                                  event_stream_port: port,
                                   max_line_captures: max_line_captures,
                                   num_lines:         program.lines.count,
                                   filename:          filename
                                 )].pack('m0')
 
-      # evaluate the code in a child process
-      opts = {
-        4 =>    child_eventstream,
-        in:     child_stdin,
-        out:    child_stdout,
-        err:    child_stderr,
-        pgroup: true, # run it in its own process group so we can SIGINT the whole group
-      }
-      child_pid  = Kernel.spawn(env, *popen_args, opts)
-      child_pgid = Process.getpgid(child_pid)
+      child = ChildProcess.build(*popen_args)
+      child.leader = true
+      child.duplex = true
+      child.environment.merge!(env)
+      child.io.stdout = child_stdout
+      child.io.stderr = child_stderr
+
+      child.start
 
       # close child streams b/c they won't emit EOF
       # until both child and parent references are closed
-      close_streams(child_eventstream, child_stdout, child_stderr, child_stdin)
-      stdin.sync = true
+      close_streams(child_stdout, child_stderr)
+      child.io.stdin.binmode
+      child.io.stdin.sync = true
```

Comment:
See the blog post again for an explanation of binary mode vs encodings. While binary mode is sometimes a way of avoiding dealing with encodings, for the most part they are two separate issues.